### ISSUE TITLE
not auto-indent  `yas-selected-text` when expand with binding for enh-ruby-mode (0.13.0 work)

### ISSUE THREAD

**Issue by zw963** [joaotavora/yasnippet#969]
Following is reproduce:

### mark `puts 'hello'` content like following

![image](https://user-images.githubusercontent.com/549126/48978658-c51f2900-f0e9-11e8-9076-bb5a7f537fe7.png)

### trigger expand with a binding, expected output is:

```rb
class TestClass
  def meth
    loop do
      puts 'hello'
    end
  end
end
```

### but, current result is:

```rb
class TestClass
  def meth
    loop do
    puts 'hello'  # <= this line not correct indent.
    end
  end
end
```

This bug is only happened for version > 0.9.1,  and with `enh-ruby-mode`,
but older version work well, so i think still some thing broken.

following is snippet definition:

```el
# -*- mode: snippet -*-
# name: loop { ... }
# key: loop
# binding: l
# condition: (yas-use-region-p)
# --
loop do
  `(yas-selected-text)`$0
end
```

**Comment by zw963** [comment]
After some research, I found this bug is introduced after 0.13.0 released, more accurate,
it introduced by 8328c38.

**Comment by npostavs** [comment]
Works for me.

> After some research, I found this bug is introduced after 0.13.0 released, more accurate,
> it introduced by [8328c38](https://github.com/joaotavora/yasnippet/commit/8328c38d77f3d6a8aafc67975437ee4fc6c09e1e).

That commit had some serious problems which should have been fixed by 1d96da2e08664c31ff7f6f7441da1f4fa5680b1f, so maybe that fixed this one as well?

**Comment by zw963** [comment]
It fixed now.

**Comment by zw963** [comment]
Hi, there still exist some issue, which introduced after 0.13.0 is released.

**Comment by zw963** [comment]
Following is a snippet example:

```cl
# -*- mode: snippet -*-
# name: do |x| structure
# key: do
# --
do${1:$(if-any " |")}${1:arg}${1:$(if-any "|")}
  $0
end
```

it use a function, which name `if-any`, defined like this:

```el
(defun if-any (&optional string1)
    (or string1 ""))
```

### expected behavior: (this work well in 0.13.0 with enh-ruby-mode)

1.  method_name => do => TAB => TAB will expand to

```rb
method_name do |x|
  I
end
```
2. method_name => do => Ctrl+d, will expand to

```rb
method_name do
  I
end
```
### Actually behavior (not worked, this bug introduced in recent version)

```rb
method_name do |x|
I
end
```
```rb
method_name do
I
end
```

**Comment by npostavs** [comment]
> method_name => do => TAB => TAB will expand to

What the "=>" mean here? Just a separator? This is what I get (assuming `I` means cursor location), I think it's as expected...

```ruby
method_name do |arg|
  I
end
```

and

```ruby
method_name do ||
  I
end
```

**Comment by npostavs** [comment]
Oops, please ignore my previous message, I was testing in an old yasnippet version.

**Comment by npostavs** [comment]
For reference, a quick reproducer, run as `emacs -Q -L .../yasnippet -L .../enh-ruby-mode -l issue-969.el`

I don't really understand why it fails.  For some reason
`enh-ruby-calculate-indent` returns 0 for the middle line when
initially expanding the snippet, but later it returns 2 as expected.
I find enh-ruby-mode's indentation code kind of hard to follow.

```emacs-lisp
(require 'enh-ruby-mode)
(require 'yasnippet)
(require 'edebug)

(with-current-buffer (car (find-function-noselect 'yas--indent))
  (edebug-instrument-function 'yas--indent)
  (forward-sexp)
  (search-backward "yas--indent-region")
  (edebug-set-breakpoint nil))

(edebug-instrument-function 'enh-ruby-calculate-indent)
(setq edebug-initial-mode 'go)


(yas-define-snippets
 'enh-ruby-mode
 '(("do" "do${1:$(or \" |\" \"\")}${1:arg}${1:$(or \"|\" \"\")}
  $0
end")))

(with-current-buffer (get-buffer-create  "*yas test*")
  (erase-buffer)
  (enh-ruby-mode)
  (yas-minor-mode)
  (insert "method do")
  (pop-to-buffer (current-buffer)))

```

**Comment by zw963** [comment]
Hi, this issue is fixed in 0.14.0, it working quite well.

Thanks for hard work!

**Comment by npostavs** [comment]
Are you sure? I don't think I've done anything to fix it, and I don't see any recent changes in enh-ruby-mode either.

**Comment by zw963** [comment]
Yes, I am quite sure about this, because i reproduce those examples again, it working.

### yas version
```sh
* 3bf9a3b - (HEAD -> master, tag: 0.14.0, upstream/release-0.14, origin/master, origin/HEAD) Release 0.14.0(3 months ago,Noam Postavsky)
```

### Emacs version

```
GNU Emacs 27.0.90 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.14, cairo version 1.17.3) of 2020-03-02
```

### my config

```cl
(require 'warnings)
(add-to-list 'warning-suppress-types '(yasnippet backquote-change))
(add-to-list 'warning-suppress-log-types '(yasnippet backquote-change))

(require 'yasnippet)
(add-to-list 'yas-snippet-dirs (expand-file-name "snippets"))
(setq  yas-also-indent-empty-lines t
       yas-also-auto-indent-first-line t
       )
(yas-global-mode t)

(add-hook 'snippet-mode-hook
          (lambda ()
            (whitespace-mode t)
            ))

(defun yas-longest-nonwordkey (start-point)
  "As `yas-key-syntaxes' element, look for longest key made of nonword characters."
  (if (= (point) start-point)
      (skip-syntax-backward "^w")
    (forward-char))
  (unless (<= start-point (1+ (point)))
    'again))
;; 查看函数 modify-syntax-entry 函数获取语法表帮助.
(setq yas-key-syntaxes '(yas-longest-nonwordkey yas-try-key-from-whitespace "w_.()" "w_." "w_"))


(provide 'yasnippet_init)
```

**Comment by npostavs** [comment]
> and I don't see any recent changes in enh-ruby-mode either.

Oh, double checked, it seems to have been fixed zenspider/enhanced-ruby-mode@3b97c6f7c4e0e462b74d57d3a0164f6b6f9b498e

**Comment by zw963** [comment]
@npostavs , Hi, a similar issue happen again on enh-ruby-mode, though it not exactly same as above.

### What i want to do.

First of all, i have to describe what i want to expand.

I want to expand a `do block` in enh-ruby-mode, like this: (TAB is tab)

```
some_code doTAB
```

will expand to (arg is $1, I is cursor)

```
some_code do |arg|
  I
end
```
But, if i press `Ctrl+D` when cursor is on arg($1)

I expect the the result will be: (I is cursor, the right/left `|` is removed too)

```
some_code do
  I
end
```

### Reproduce.

my snippet

```
# -*- mode: snippet -*-
# name: do |x| structure
# key: do
# --
do${1:$(if-any " |")}${1:arg}${1:$(if-any "|")}
  $0
end
```

because above snippet use `(if-any)` function, following is it definition.

```
(defun if-any (&optional string1 string2)
  "Add STRING1 if `yas-text' content is not empty, otherwise STRING2."
  (if (string-match "^$" (or yas-text yas-selected-text))
      string2
    (or string1 "")
    )
  )
```

### the Issue

The above snippet. work quite well with 0.13, but not work well with 0.14.

Because the expanded result indent error, as before: (I is cursor)

if not press Ctrl+D

```
some_code do |arg|
I
end
```

if press Ctrl + D

```
some_code do
I
end
```

BTW:

1.  previous hack on zenspider/enhanced-ruby-mode@3b97c6f was seem like reverted, but,  original issue in this issue seem like still work.
2. this issue can not be reproduced on `ruby-mode`, it only happen on enh-ruby-mode


Thank you.

**Comment by zenspider** [comment]
Because I reverted it via https://github.com/zenspider/enhanced-ruby-mode/commit/2cacb31a8b766c36840f88c7e88074471b1fca83

`(erm-wait-for-parse)` in the right place will probably get you the right indent values, but it isn't going to be in ERM's `before-change-functions` hook

**Comment by zw963** [comment]
> Because I reverted it via [zenspider/enhanced-ruby-mode@2cacb31](https://github.com/zenspider/enhanced-ruby-mode/commit/2cacb31a8b766c36840f88c7e88074471b1fca83)
> 
> `(erm-wait-for-parse)` in the right place will probably get you the right indent values, but it isn't going to be in ERM's `before-change-functions` hook

Okay, thank you, i will keeping use old v0.13.0 to workaround this.
